[REVIEW] Fix rolling-window count for null input

mythrocks opened the pull request:
Closes #6343.

Fixes `COUNT_ALL`, `COUNT_VALID` for window functions.

In rolling_window() operations, COUNT_VALID/COUNT_ALL should only return null rows if the min_periods requirement is not satisfied. For all other cases, the count produced must be valid, even if the input row is null.

As it currently stands, the COUNT* rolling_window() operation returns null if even one of its input rows is null. That behaviour, while correct for aggregations like SUM, is incorrect for COUNT.

This commit should fix the problem.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

revans2 commented:
This looks like it would fix the issue for Spark where we cannot do a COUNT on a column for a rolling window.

@kkraus14 is this going to mess anything up for python side

mythrocks commented:
(Apologies for the force-push. I thought I'd get the fix for the code-break in before the review begins.)

mythrocks commented:
> @kkraus14 is this going to mess anything up for python side

A more complete accounting of the problem is available at #6343, in case the description here is too terse.

kkraus14 commented:
> @kkraus14 is this going to mess anything up for python side

cc @shwina @brandon-b-miller who worked on rolling windows

brandon-b-miller commented:
I don't think we rely on this operation producing a particular result internally for anything on the python side, so I don't _think_ anything will end up happening except for the associated changes propagating to the user facing version of this API. 

I think this might address https://github.com/rapidsai/cudf/issues/5580 and is related to pandas changes discussed in  https://github.com/pandas-dev/pandas/issues/34466 and the issues linked to therein. Honestly, when @shwina and I looked at this a few months back, we had a hard time parsing out the logic behind the pandas behavior. It seemed like no matter how we preprocessed `min_periods` and `win_size` with count, there was always a set of cases where we got a different answer than pandas and it was very hard to pin down what was happening. 

I'm curious to see what happens if we build this branch and remove the `xfails` which mark those tests right now - I'll build this branch this afternoon and try it and report back here.

mythrocks commented:
> I think this might address #5580 and is related to pandas changes discussed in pandas-dev/pandas#34466

Thank you for the link, @brandon-b-miller. Your assessment is accurate. 
As a side-note, was anyone else confused with pandas-dev's nomenclature regarding "the row _above_"?

>> When your window size is an even number, with `center=True` you actually include 1 more element above the anchor than below the anchor.

(He means the preceding row, right? This was an informative discussion.) Regardless, it appears that behaviour of count in `pandas < 1.0` was incorrect as well.

mythrocks commented:
Some of the python tests seem to be failing. 

```
=========================== short test summary info ============================
FAILED cudf/tests/test_rolling.py::test_rolling_with_offset[count] - Assertio...
= 1 failed, 53475 passed, 2206 skipped, 954 xfailed, 4040 xpassed, 55395 warnings in 3978.62s (1:06:18) =
```

Is there an easy way to run a specific Python test? I suspect these tests will need their expected output modified.

brandon-b-miller commented:
You can use `py.test test_rolling.py -k test_rolling_with_offset[count]` to run just that one test

mythrocks commented:
The failure is quite interesting. Reproduced here in full, to save on browser tabs:
```
agg = 'count'
    @pytest.mark.parametrize(
        "agg", ["sum", pytest.param("min"), pytest.param("max"), "mean", "count"]
    )
    def test_rolling_with_offset(agg):
        psr = pd.Series(
            [1, 2, 4, 4, np.nan, 9],
            index=[
                pd.Timestamp("20190101 09:00:00"),
                pd.Timestamp("20190101 09:00:01"),
                pd.Timestamp("20190101 09:00:02"),
                pd.Timestamp("20190101 09:00:04"),
                pd.Timestamp("20190101 09:00:07"),
                pd.Timestamp("20190101 09:00:08"),
            ],
        )
        gsr = cudf.from_pandas(psr)
>       assert_eq(
            getattr(psr.rolling("2s"), agg)().fillna(-1),
            getattr(gsr.rolling("2s"), agg)().fillna(-1),
            check_dtype=False,
        )

cudf/tests/test_rolling.py:125: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/_libs/testing.pyx:67: in pandas._libs.testing.assert_almost_equal
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   AssertionError: Series are different
E   
E   Series values are different (16.66667 %)
E   [index]: [2019-01-01T09:00:00.000000000, 2019-01-01T09:00:01.000000000, 2019-01-01T09:00:02.000000000, 2019-01-01T09:00:04.000000000, 2019-01-01T09:00:07.000000000, 2019-01-01T09:00:08.000000000]
E   [left]:  [1.0, 2.0, 2.0, 1.0, -1.0, 1.0]
E   [right]: [1, 2, 2, 1, 0, 1]
```

For some odd reason, the fifth input causes a `nan` to be produced, (and replaced with `-1.0`). I'm poring through the `pandas` documentation now, to see if I can make sense of this. On the face of it, it would appear that in `pandas`, the `nan` rows cause a `nan` output, but only in the indexed column case.

mythrocks commented:
I suspect that the Pandas [series rolling-count `min_periods` fix](https://github.com/pandas-dev/pandas/pull/30923) has not been extended to "indexed" columns (using time-range based windows). I have [requested the author for clarification](https://github.com/pandas-dev/pandas/issues/34466#issuecomment-701236207) on this behaviour.

If my hunch is right, we might need to move this test to `xfail` for the moment.

brandon-b-miller commented:
There's some tests in `test_rolling.py` that are marked as `xfail` right now due to these issues. I tried removing the marks, but I'm still seeing failures. Here's an example. Any idea what might be happening?

```
>>> import pandas as pd
>>> import cudf
>>> psr = pd.Series([1,1,1,1])
>>> gsr = cudf.Series([1,1,1,1])
>>> psr.rolling(2,2,True).count()
0    NaN
1    2.0
2    2.0
3    2.0
dtype: float64
>>> gsr.rolling(2,2,True).count()
0    1
1    2
2    2
3    2
dtype: int32
>>> pd.__version__
'1.1.2'
```

mythrocks commented:
> Any idea what might be happening?

On the face of it, my code seems to be producing the incorrect result. I will investigate.

brandon-b-miller commented:
Connected offline and found this to be part of the issue https://github.com/rapidsai/cudf/blob/branch-0.16/python/cudf/cudf/core/window/rolling.py#L233-L234

mythrocks commented:
@brandon-b-miller is right, of course. `cudf::rolling_window()` seems to be working correctly, as tested with a temporarily added test-case:
```
42: [ RUN      ] UntypedRollingTest.TestMinPeriods
42: Input vector: 1,1,1,1
42: count(preceding=2, following=0, min_periods=2) : NULL,2,2,2
42: count(preceding=2, following=0, min_periods=1) : 1,2,2,2
42: count(preceding=2, following=0, min_periods=0) : 1,2,2,2
42: [       OK ] UntypedRollingTest.TestMinPeriods (0 ms)
```

I'll post a correction for the Python tests.

mythrocks commented:
Drat. Removing the `min_periods` hack from `rolling.py` breaks `test_rolling_groupby_*`. This will need a more delicate hand.

brandon-b-miller commented:
I'll look into this today.

mythrocks commented:
> I'll look into this today.

I'm on this as well. I'll ping you as soon as I have something.

galipremsagar commented on python/cudf/cudf/tests/test_rolling.py:
```diff
@@ -122,11 +108,19 @@ def test_rolling_with_offset(agg):
+            check_dtype=False,
+        )
+    except AssertionError as e:
+        if agg == "count": 
+            # count for offset-columns does not follow similar conventions
+            # to non-offset-columns.
+            pytest.xfail(reason="Differs from pandas behaviour here") 
+        else:
+            raise e
```
Could we mark `"count"` parameter similar to this: https://github.com/rapidsai/cudf/blob/branch-0.16/python/cudf/cudf/tests/test_dataframe.py#L2092-#L2097

```python
pytest.param(
            "count",
            marks=pytest.mark.xfail(
                reason="Differs from pandas behaviour here"
            ),
        )
```

mythrocks replied:
Ah, right you are. Apologies, I was following the convention in `test_rolling.py`. I'll switch.

mythrocks replied:
Thanks for the suggestion, @galipremsagar. This is cleaner.

mythrocks commented:
@brandon-b-miller, I have moved the erstwhile hack for `COUNT` from `Rolling` to `RollingGroupby`, in the latest commit.

I have also removed the `xfail` tests. They appear to be passing for me. I'll wait to see how the CI fares.

Let me take another shot at [the discussion on the Pandas issue](https://github.com/pandas-dev/pandas/issues/34466#issuecomment-701236207).

mythrocks commented:
@vuule, @trxcllnt: Could I bother you to please look at the C++-side change?

vuule commented on cpp/src/rolling/rolling.cu:
```diff
@@ -70,10 +70,42 @@ process_rolling_window(column_device_view input,
+                       size_type current_index,
+                       size_type min_periods)
+{
+  // declare this as volatile to avoid some compiler optimizations that lead to incorrect results
+  // for CUDA 10.0 and below (fixed in CUDA 10.1)
+  volatile cudf::size_type count = end_index - start_index;
```
Can you please try with 
```suggestion
  cudf::size_type const count = end_index - start_index;
```
in this implementation? the CUDA bug is unlikely to affect this version of the kernel too.

mythrocks replied:
Good point. Removed.

vuule reviewed (changes requested):
Changes make sense. Just a small question+suggestion below.

mythrocks commented on cpp/src/rolling/rolling.cu:
```diff
@@ -70,10 +70,42 @@ process_rolling_window(column_device_view input,
   }
 
+  bool output_is_valid                      = ((end_index - start_index) >= min_periods);
+  output.element<OutputType>(current_index) = count;
```
Good catch. I shouldn't be setting this.

vuule replied:
In that case, I suggest returning early and looping only when the output is valid :)
It's possible now since we don't use the count to determine validity.

mythrocks replied:
Yep, that's a good one. Fixed.

mythrocks commented:
@vuule, thank you for the review and suggestions for improvement. :]

mythrocks commented:
@trxcllnt,

Argh, I hadn't realized that this PR was still tagged `[WIP]`, not `[REVIEW]`. I have corrected this. Apologies if this review slipped off your radar.

harrism commented:
Suggest using draft PRs for WIP and dropping the WIP/REVIEW annotations in the future.

harrism commented on cpp/src/rolling/rolling.cu:
```diff
@@ -58,7 +58,7 @@ template <typename InputType,
           aggregation::Kind op,
           bool has_nulls>
-std::enable_if_t<op == aggregation::COUNT_VALID || op == aggregation::COUNT_ALL, bool> __device__
+std::enable_if_t<op == aggregation::COUNT_VALID, bool> __device__
```
Is there a reason to use return-type-SFINAE over template-parameter-SFINAE?

mythrocks replied:
No, sir, except to retain the erstwhile  convention for `process_rolling_window()`. I would prefer the template parameter, myself.
Would you recommend changing this out?

harrism replied:
If it's not a big deal. We use template parameter SFINAE almost everywhere.

mythrocks replied:
> If it's not a big deal.

Hardly any trouble. I've switched this to template-parameter based SFINAE.

harrism commented on cpp/src/rolling/rolling.cu:
```diff
@@ -70,10 +70,42 @@ process_rolling_window(column_device_view input,
+    if (!has_nulls) {
+      count = end_index - start_index;
+    } else {
+      for (size_type j = start_index; j < end_index; j++) {
+        if (input.is_valid(j)) { count++; }
```
Could use a `thrust::count_if(thrust::seq` here?

mythrocks replied:
Changed to `thrust::count_if()`.

mythrocks replied:
P.S. Thank you for the suggestion.

harrism commented on python/cudf/cudf/tests/test_rolling.py:
```diff
@@ -47,15 +47,7 @@ def test_rollling_series_basic(data, index, agg, nulls, center):
             got = getattr(
                 gsr.rolling(window_size, min_periods, center), agg
             )().fillna(-1)
-            try:
```
Please fix the spelling of the function name (can't comment on it, but it is `test_rollling_series_basic` (3 letter `l`).

mythrocks replied:
Yes, that bothered me. Will fix.

harrism reviewed (changes requested):
Looks good. Couple small things.  My review can replace @trxcllnt 's if  he is busy.

mythrocks commented:
@harrism, thank you for your review and suggestions. Please let me know if this PR is looking alright to go in. (I'm a little nervous about rebasing #6277 in time for tomorrow.)

harrism commented on cpp/src/rolling/rolling.cu:
```diff
@@ -76,9 +78,10 @@ process_rolling_window(column_device_view input,
+      count = thrust::count_if(thrust::seq,
+                               thrust::make_counting_iterator(start_index),
+                               thrust::make_counting_iterator(end_index),
+                               [&input](auto i) { return input.is_valid(i); });
```
To avoid the `is_nullable()` check on every access to is_valid...

```suggestion
                               [&input](auto i) { return input.is_valid_no_check(i); });
```

mythrocks replied:
Thanks, I had to look that up. `has_nulls` is actually a one-time reading of `input.has_nulls()`, so it should be safe to call `is_valid_no_check()`.
Super. I'll change this and test it.

mythrocks commented:
rerun tests

mythrocks commented:
@shwina, might I please bug you to have a look at the Python bits, at your earliest convenience?

mythrocks commented:
Thanks for the review, @harrism!

shwina commented on python/cudf/cudf/tests/test_rolling.py:
```diff
@@ -97,17 +89,24 @@ def test_rolling_dataframe_basic(data, agg, nulls, center):
+        pytest.param("max"),
+        "mean",
+        pytest.param(
+            "count",  # Does not follow similar conventions as
```
Since we are removing `count` from this test because it does't match Pandas behaviour, should we add another test for `count` that checks against what _we_ consider to be correct behaviour?

mythrocks replied:
I've added an explicit test for offset rolling windows, for count.

kkraus14 commented on python/cudf/cudf/tests/test_rolling.py:
```diff
@@ -97,17 +89,24 @@ def test_rolling_dataframe_basic(data, agg, nulls, center):
-    "agg", ["sum", pytest.param("min"), pytest.param("max"), "mean", "count"]
+    "agg",
+    [
+        "sum",
+        pytest.param("min"),
+        pytest.param("max"),
+        "mean",
```
Can we make these consistent and just make them all strings or all `pytest.param`

mythrocks replied:
Agreed. They're all `pytest.param".

shwina reviewed (approved):
LGTM. Thanks @mythrocks!

mythrocks commented:
Thanks for the review, @shwina (and everyone). Thanks for merging this, @harrism!
